src/messaging/ExchangeContext.cpp

```diff
@@ -213,17 +214,29 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
         else
         {
 #endif
-            err = mDispatch.SendMessage(GetExchangeMgr()->GetSessionManager(), mSession.Get().Value(), mExchangeId, IsInitiator(),
+            err = mDispatch.SendMessage(GetExchangeMgr()->GetSessionManager(), session, mExchangeId, IsInitiator(),
                                         GetReliableMessageContext(), reliableTransmissionRequested, protocolId, msgType,
                                         std::move(msgBuf));
 #if CONFIG_BUILD_FOR_HOST_UNIT_TEST
         }
 #endif
-        // We should only cancel the response timer if the ExchangeContext fails to send the message that starts the response timer.
-        if (err != CHIP_NO_ERROR && startedResponseTimer)
+        if (err != CHIP_NO_ERROR)
         {
-            CancelResponseTimer();
-            SetResponseExpected(false);
+            // We should only cancel the response timer if the ExchangeContext fails to send the message that starts the response
+            // timer.
+            if (startedResponseTimer)
+            {
+                CancelResponseTimer();
+                SetResponseExpected(false);
+            }
+
+            // If we can't even send a message (send failed with a non-transient
+            // error), mark the session as defunct, just like we would if we
+            // thought we sent the message and never got a response.
+            if (session->IsSecureSession() && session->AsSecureSession()->IsCASESession())
+            {
+                session->AsSecureSession()->MarkAsDefunct();
```

Comment:
I wonder if marking the session as defunct isn't overreacting a bit :). What we may suspect in such a case is that the destination address is no longer valid, but we have no premise that the message has been rejected by the destination due to the usage of an old session. So maybe we should rather only try to update the address. On the other hand, then we would need some throttling mechanism to limit the mDNS queries...

Reply:
I considered just updating the address, in fact.  But the machinery for doing that has been more or less ripped out at this point, so it's _quite_ difficult to do.  But yes, if it would be preferable I can resurrect that machinery and use it here....